docs/DesignDocs/SealingAPI.md

```diff
@@ -0,0 +1,591 @@
+  contains SGX specific definitions.
+- [enclave/seal_gcm-aes.c](../../enclave/seal_gcm-aes.c) implements `oe_seal()`
+  and `oe_unseal()`.
+- [enclave/crypto/gcm.c](../../enclave/crypto/gcm.c) implements GCM-AES128.
+
+Modified files:
+
+- [include/openenclave/enclave.h](../../include/openenclave/enclave.h)
+  declares `oe_get_seal_key_info()`, `oe_seal()` and `oe_unseal()`.
+- [enclave/core/sgx/keys.c](../../enclave/core/sgx/keys.c) implements
+  `oe_get_seal_key_info()`.
+
+### oe_seal()
+
+```C
+/**
+ * Seal data to an enclave using AEAD (Authenticated Encryption with Additional
+ * Data).
+ *
+ * @param[in] key_info The enclave-specific key information to derive the seal
+ * key with.
+ * @param[in] key_info_size The size of the \p key_info buffer.
+ * @param[in] plaintext Optional buffer to be encrypted under the seal key.
+ * @param[in] plaintext_size Size of \p plaintext, must be \c 0 if \p plaintext
+ * is \c NULL.
+ * @param[in] additional_data Optional additional data to be included in the
+ * final MAC.
+ * @param[in] additional_data_size Size of \p additional_data, must be \c 0 if
+ * \p additional_data is \c NULL.
+ * @param[out] blob On success, receive the pointer to a buffer containing both
+ * \p additional_data and encrypted \p plaintext, along with necessary
+ * information for unsealing. Freed by oe_free().
+ * @param[out] blob_size On success, receive the size of \p blob.
+ *
+ * @retval OE_OK \p plaintext and \p additional_data were sealed to the enclave
+ * successfully.
+ * @retval OE_INVALID_PARAMETER At least one parameter is invalid.
+ * @retval OE_OUT_OF_MEMORY Failed to allocate memory.
+ * @retval OE_CRYPTO_ERROR Error occurred during encryption.
+ */
+oe_result_t oe_seal(
+    const uint8_t* key_info,
+    size_t key_info_size,
+    const uint8_t* plaintext,
+    size_t plaintext_size,
+    const uint8_t* additional_data,
+    size_t additional_data_size,
+    uint8_t **blob,
+    size_t *blob_size);
+```
+
+`oe_seal()` encrypts and packages user data into a sealed blob using *AEAD*
+(*Authenticated Encryption with Additional Data*). A sealed blob is comprised
+of a header, followed immediately by payload, which is the concatenation of
+cipher text and *AAD* (*Additional Authenticated Data*).
+
+The blob header is TEE specific. On SGX, for interoperability, we adopted the
+same format as the Intel SGX SDK. Below gives the definition of
+`oe_sealed_blob_header_t` on SGX.
+
+```C
+typedef struct _sgx_sealed_blob_header
+{
+    sgx_key_request_t key_info;
+    uint32_t ciphertext_size; /* also offset of AAD into payload, as cipher
+                                 text is followed immediately by AAD */
+    uint8_t reserved[12];     /* must be 0 */
+    uint32_t payload_size;    /* total size of cipher text and AAD */
+    uint8_t iv[12];  /* must be 0 to be compatible with Intel SGX SDK */
+    uint8_t tag[16]; /* Authentication Tag */
+} oe_sealed_blob_header_t;
+
+#define ki_entropy key_id
```

Comment:
It is to make `oe_seal()`/`oe_unseal()` source code compatible with multiple TEEs. 

Different TEEs use different structure fields for holding user supplied entropy. Here a more generic name `ki_entropy` (`ki` stands for `key_info`) is introduced, so that `oe_seal()` can always use `key_info->ki_entropy` to reference the correct structure member. On SGX, it's `sgx_key_request_t::key_id`.